src/charm.py

```diff
@@ -100,19 +107,34 @@ def container(self) -> Container:
         container = self.unit.get_container(FLASK_CONTAINER_NAME)
         return container
 
-    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
-        """Configure the flask pebble service layer.
+    @property
+    def _is_precondition_satisfied(self) -> bool:
+        """Check if the precondition for the Flask application has been satisfied.
 
-        Args:
-            event: the config-changed event that trigger this callback function.
+        Preconditions include:
+            1. Flask workload container is ready
+            2. Secret storage has been initialized
+
+        Return:
+            True if all preconditions is satisfied.
         """
         try:
-            container = self.container()
+            self.container()
         except PebbleNotReadyError:
-            logger.info("pebble client in the Flask container is not ready, defer config-changed")
-            event.defer()
-            return
-        container.add_layer("flask-app", self.flask_layer(), combine=True)
+            logger.info("pebble client in the Flask container is not ready")
+            return False
+        if not self._secret_storage.is_initialized:
+            logger.info("secret storage is not initialized, defer config-changed")
+            return False
+        return True
+
+    def _restart_flask_application(self) -> None:
+        """Start or restart the flask application with the latest charm configuration.
+
+        The flask charm must be ready (i.e. preconditions meet) before calling this function.
```

Comment:
If the preconditions must be met, then we should check them in this method and raise an exception if they're not met

Reply:
I'm not sure about that. From what I can see, not meeting those preconditions would be part of the normal flow, for instance, when pebble is not ready

Reply:
If the preconditions aren't met, the charm should defer the event. But the `_restart_flask_application` doesn't have any context of the event, so the check need to be done outside.